whois-update/src/main/java/net/ripe/db/whois/update/handler/validator/keycert/X509KeycertValidator.java

```diff
@@ -0,0 +1,101 @@
+package net.ripe.db.whois.update.handler.validator.keycert;
+
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Sets;
+import net.ripe.db.whois.common.DateTimeProvider;
+import net.ripe.db.whois.common.Message;
+import net.ripe.db.whois.common.Messages;
+import net.ripe.db.whois.common.domain.CIString;
+import net.ripe.db.whois.common.rpsl.AttributeType;
+import net.ripe.db.whois.common.rpsl.ObjectType;
+import net.ripe.db.whois.common.rpsl.RpslObject;
+import net.ripe.db.whois.update.domain.Action;
+import net.ripe.db.whois.update.domain.PreparedUpdate;
+import net.ripe.db.whois.update.domain.UpdateContext;
+import net.ripe.db.whois.update.domain.UpdateMessages;
+import net.ripe.db.whois.update.handler.validator.BusinessRuleValidator;
+import net.ripe.db.whois.update.keycert.X509CertificateWrapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.security.PublicKey;
+import java.security.interfaces.DSAPublicKey;
+import java.security.interfaces.RSAPublicKey;
+import java.util.Set;
+
+@Component
+public class X509KeycertValidator implements BusinessRuleValidator {
+
+    private static final ImmutableList<Action> ACTIONS = ImmutableList.of(Action.CREATE);
+    private static final ImmutableList<ObjectType> TYPES = ImmutableList.of(ObjectType.KEY_CERT);
+
+    private static final CIString METHOD_X509 = CIString.ciString("X509");
+
+    private static final Set<String> WEAK_HASH_ALGORITHMS = Sets.newHashSet("SHA1withRSA", "SHA1withDSA", "MD5withRSA", "MD5withDSA");
```

Comment:
So md2WithRSAEncryption (see https://www.ietf.org/rfc/rfc3279.txt) would be ok?
I think it's better to whitelist here or even check just for hash algorithm presence (so should be SHA256, SHA384, SHA512) and make those configurable.

Reply:
I prefer a blocklist for known weak hash algorithms, rather than blocking new (more secure) algorithms by default.

I generated this list using the known keys in the key-cert split file, I can always add more.

We could indeed extract this to a property (with a default value), so we can keep it updated without needing a release, but the list changes very rarely.

I'm doing a string comparison because java.security.cert.X509Certificate seems to only return a string (either getSigAlgName() or getSigAlgOID()). There doesn't appear to be an integer value for the X509 hash algorithm, like in org.bouncycastle.bcpg.PublicKeyAlgorithmTags which I can use for PGP keys.